### PR TITLE
Add eigen gru

### PR DESCRIPTION
**Pull request by joey12300**
### PR types

Performance optimization
### PR changes

OPs
### Describe

* fix dropout backward bug
* optimize gru backward
* Add eigen gru

Here is performance of GRU among paddle, pytorch and tensorflow. We run the performance script **50** times to  calculate the average forward, backward time per experiment. Following is the experiment setting:
* batch_size: 64
* time_step: 32
* epoch: 200
* input_size: 128
* hidden_size: 128
* num_layers: 1
* dropout: 0.0

Performance data for eigen **GRU**:
```
forward:
paddle:         
mean: 0.01042 s  std: 0.00157
pytorch: 
mean: 0.01320 s  std: 0.00161
tensorflow:  
mean: 0.06519 s std: 0.00224

Compared to Pytorch and TensorFlow, the forward speed increased by 26.67%, 525.62% respectively.

backward:
paddle:
mean: 0.01270 s std: 0.00217
pytorch:
mean: 0.02157 s std: 0.00256
tensorflow:
mean: 0.06153 s std: 0.00217

Compared to Pytorch and TensorFlow, the backward speed increased by 69.84%, 384.49% respectively.
```

Supplement  performance data for eigen **LSTM**:
```
forward:
paddle:         
mean: 0.01326 s  std: 0.00266
pytorch: 
mean: 0.01637 s  std: 0.00309
tensorflow:  
mean: 0.06102 s std: 0.00333

Compared to Pytorch and TensorFlow, the forward speed increased by 23.45%,  360.18% respectively.

backward:
paddle:
mean: 0.01809 s std: 0.00372
pytorch:
mean:0.02630 s std:0.00481
tensorflow:
mean:0.05803 s std:0.00366

Compared to Pytorch and TensorFlow, the backward speed increased by 45.38%,  220.78% respectively.
```

### CODE REVIEW

File: paddle/fluid/operators/math/detail/gru_cpu_kernel.h [hunk 1]
```diff
@@ -268,25 +299,52 @@ inline void forward_reset_output(OpResetOutput op_reset_output,
...
-      hl_avx_gru_forward_final_output(op_final_output, value.gate_value,
-                                      value.prev_out_value, value.output_value,
-                                      frame_size, active_node, origin_mode,
-                                      old_version);
+    if (!old_version) {
+      // eigen
+      eigen_forward_final_output(*context, value, frame_size);
     } else {
-      hl_naive_gru_forward_final_output(op_final_output, value.gate_value,
+      if (OpFinalOutput::avx && (frame_size > static_cast<int>(8 - 1)) &&
```
Comment (ZeyuChen) [inline paddle/fluid/operators/math/detail/gru_cpu_kernel.h]: 这个判断里面又hard coding 8-1是为了？

File: paddle/fluid/operators/rnn_op.h [hunk 2]
```diff
@@ -210,66 +210,58 @@ struct LSTMCell : Cell<T> {
   }
 };
 
+template <typename T>
+void dropout_helper(const framework::ExecutionContext& context, Tensor* x,
+                    Tensor* y, const Tensor* mask, const float& dropout_prob) {
+  auto& place = *context.template device_context<platform::CPUDeviceContext>()
+                     .eigen_device();
+  auto eigen_dropout_mask = EigenVector<uint8_t>::Flatten(*mask);
```
Comment (ZeyuChen) [inline paddle/fluid/operators/rnn_op.h]: 我觉得整体的代码不用将eigen作为前缀来描述变量。一方面冗余，第二方面没必要，就是个数据计算库，没什么特殊需要对变量进行前缀引入

File: paddle/fluid/operators/rnn_op.h [hunk 3]
```diff
@@ -210,66 +210,58 @@ struct LSTMCell : Cell<T> {
...
+  auto eigen_x = EigenVector<T>::Flatten(*x);
+  auto eigen_y = EigenVector<T>::Flatten(*y);
+  if (dropout_prob == 1.0f) {
+    eigen_y.device(place) = static_cast<T>(0) * eigen_x;
```
Comment (ZeyuChen) [inline paddle/fluid/operators/rnn_op.h]: 就用普通的x和y来描述如何？

File: paddle/fluid/operators/math/detail/gru_cpu_kernel.h [hunk 4]
```diff
@@ -664,23 +722,74 @@ inline void backward_reset_grad(OpResetGrad op_reset_grad,
   }
 }
 
+template <typename T>
+inline void eigen_gru_backward(const platform::CPUDeviceContext &context,
+                               GRUMetaValue<T> value, GRUMetaGrad<T> grad,
+                               int frame_size) {
+  auto &place = *context.eigen_device();
+
+  auto eigen_value_reset_gate =
```
Comment (ZeyuChen) [inline paddle/fluid/operators/math/detail/gru_cpu_kernel.h]: 是否可以去掉eigen的变量名前缀？

File: paddle/fluid/operators/math/detail/gru_cpu_kernel.h [hunk 5]
```diff
@@ -242,23 +248,48 @@ void hl_avx_gru_forward_final_output(OpFinalOutput op_final_output,
 #endif
 }
 
+template <typename T>
+inline void eigen_forward_reset_output(
+    const platform::CPUDeviceContext &context, GRUMetaValue<T> value,
+    int frame_size) {
+  auto &place = *context.eigen_device();
+  auto eigen_value_reset_gate =
```
Comment (ZeyuChen) [inline paddle/fluid/operators/math/detail/gru_cpu_kernel.h]: 再变量名这里弱化下eigen吧

**Review by ZeyuChen — CHANGES_REQUESTED** [review]
整体的函数接口和函数名，变量名弱化下eigen。Eigen只是个底层计算库，不用再各种地方突显。这样会对将来计算库更换造成很大的修改成本。

File: paddle/fluid/operators/math/detail/gru_cpu_kernel.h (hunk 1 above)
Comment (joey12300) [reply paddle/fluid/operators/math/detail/gru_cpu_kernel.h]: 这里的代码应该写错了，应该是`frame_size & (8 - 1)`，为了用位操作判断frame_size是否为8的倍数,这是使用avx需要满足的条件。

File: paddle/fluid/operators/math/detail/gru_cpu_kernel.h (hunk 4 above)
Comment (joey12300) [reply paddle/fluid/operators/math/detail/gru_cpu_kernel.h]: 已经将所有涉及eigen前缀的函数、变量删除eigen前缀

File: paddle/fluid/operators/math/detail/gru_cpu_kernel.h [hunk 6]
```diff
@@ -664,23 +719,70 @@ inline void backward_reset_grad(OpResetGrad op_reset_grad,
...
-          grad.prev_out_grad, value.reset_output_value, grad.reset_output_grad,
-          grad.output_grad, frame_size, active_node, active_gate);
-    } else {
-      hl_naive_gru_backward(
-          op_gru_grad, value.gate_value, grad.gate_grad, value.prev_out_value,
-          grad.prev_out_grad, value.reset_output_value, grad.reset_output_grad,
-          grad.output_grad, frame_size, active_node, active_gate);
-    }
+    // eigen
+    gru_backward(context, value, grad, frame_size);
```
Comment (wawltor) [inline paddle/fluid/operators/math/detail/gru_cpu_kernel.h:785]: 这块直接把之前的gru_backward会不会有影响？ 因为这块的测速结论还没有真正在inference确定

**Review by wawltor — APPROVED** [review]
LGTM

File: paddle/fluid/operators/math/detail/gru_cpu_kernel.h (hunk 6 above)
Comment (joey12300) [reply paddle/fluid/operators/math/detail/gru_cpu_kernel.h:785]: 这是全新的函数，把avx去掉不影响以前版本GRU的使用

**Comment by ZeyuChen** [comment]
LGTM

**Review by ZeyuChen — APPROVED** [review]
LGTM
